### PR TITLE
[CSM-169] Added function(s) for iterating all blocks matching predicate.

### PR DESCRIPTION
**Pull request by ksaric**
Added functions that support the (naive) way of fetching blocks by epoch/slot proposed by @gromakovsky.

### CODE REVIEW

File: src/Pos/DB/Block.hs [hunk 1]
```diff
@@ -102,6 +103,31 @@ loadDataWhile getter predicate start = NewestFirst <$> doIt start
                 then (d :) <$> doIt prev
                 else pure []
 
+-- Return all blocks that match the given predicate. It is repetetive,
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: If function is exported, then better use haddock.

File: src/Pos/DB/Block.hs [hunk 2]
```diff
@@ -102,6 +103,31 @@ loadDataWhile getter predicate start = NewestFirst <$> doIt start
...
+    => (HeaderHash -> m a)
+    -> (a -> Bool)
+    -> HeaderHash
+    -> m (NewestFirst [] a)
+loadDataMatching getter predicate start = NewestFirst <$> doIt start []
+  where
+    doIt :: HeaderHash -> [HeaderHash] -> m [a]
+    doIt h acc
+        -- When we reach the genesis block, return the accumulator.
+        | h == genesisHash = sequence $ getter <$> acc
```
Comment (gromakovsky) [inline src/Pos/DB/Block.hs]: It's inefficient to use `getter` twice, because it accesses DB and it can be slow. Why not store `[a]` in accumulator?

**Comment by gromakovsky** [comment]
> ksaric wants to merge 1 commit into cardano-sl-0.3 from feature/csm-169-extend-search-epoch
Why `cardano-sl-0.3`? AFAIU, it should go to `cardano-sl-0.3-mid` branch, but for some reason we don't have it. @georgeee why?

File: src/Pos/DB/Block.hs (hunk 2 above)
Comment (ksaric) [inline src/Pos/DB/Block.hs]: Nice. Didn't even notice that.

File: src/Pos/DB/Block.hs (hunk 1 above)
Comment (ksaric) [inline src/Pos/DB/Block.hs]: Ok, will fix.

**Comment by akegalj** [comment]
I agree with @gromakovsky . This should go to `-mid` branch

@ksaric can you make PR to mid branch. I will review and merge it. 

After this create CSM task to "Optimize epoch/slot query" or something, add details about this task (link to this PR and links from slack's channel where gromak explained stuff) and assign it to @flyingleafe

File: src/Pos/DB/Block.hs [hunk 3]
```diff
@@ -102,6 +103,31 @@ loadDataWhile getter predicate start = NewestFirst <$> doIt start
                 then (d :) <$> doIt prev
```
Comment (akegalj) [inline src/Pos/DB/Block.hs]: `doIt` is kind of poor naming?

Comment (ksaric) [inline src/Pos/DB/Block.hs]: That name was already used in the similar function named `loadDataWhile`, I didn't want to change existing code. But, I agree, `go` would be more idiomatic.

**Pull request closed without merging**